---
title: "Genome-wide association analysis, data `r params$group`, disease model: `r params$disease.model`"
author:
  - "trnguyen@ukaachen.de"
date: "Last update on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: no
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 3
    number_sections: true
    theme: lumen
params:
  group: NA
  maf: NA
  disease.model: NA
---


```{css zoom-lib-src, echo = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```

```{js zoom-jquery, echo = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>

# Code description 

<!-- Please skip this part if you are not going to re-run the scripts.  -->

<!-- parallel -j 20 Rscript 01_generate_ped_and_fam_files.R -i {} -o '/home/hieunguyen/CRC1382/outdir/AGBruns_SNP_data_full/OUTPUT/parallel_processing/output_chunks' ::: /home/hieunguyen/CRC1382/outdir/AGBruns_SNP_data_full/OUTPUT/parallel_processing/input_chunks/* -->


Raw data were transformed to `.ped` format by custom R-script and BASH-script.

- **with disease/condition**: affected = 1 

- **withut disease/condition**: affected = 0

**Note**: To control the False discovery rate, we usually select variants whose GWAS p-value is less than 5e-8. 


```{r echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, results='hide'}
#####----------------------------------------------------------------------#####
# PREPARATION AND CONFIGURATIONS
#####----------------------------------------------------------------------#####
# gc()
# rm(list = ls())

path.to.project.src <- "/home/hieunguyen/CRC1382/src_2023/AGBruns_SNP_data_full"

source(file.path(path.to.project.src, "00_import_libraries.R"))
source(file.path(path.to.project.src, "00_helper_functions.R"))
library(karyoploteR)
##### Notes
# - RAF: reference allele frequency.
# - MAF: minor allele frequency.
```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, results='asis'}
##### QC filtering thresholds
group <- params$group
maf <- params$maf
disease.model <- params$disease.model

# group <- "SBP_ever"
# maf <- 0.01
# disease.model <- "DOM"

num.show.topSNPs <- 100
thres_call.freq <- 0.5
thres_maf <- maf
thres_hwe <- 0.05

print("QC criteria: ")
print(sprintf("Call frequency >= %s", thres_call.freq))
print(sprintf("HWE Chi square test p.val >= %s", thres_hwe))
print(sprintf("MAF >= %s", thres_maf))

outdir <- "/home/hieunguyen/CRC1382/outdir"
PROJECT <- "AGBruns_SNP_data_full_corrected_20240312"
```

```{r echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, results='asis'}
path.to.save.RData <- file.path(outdir, PROJECT, "snpStats_RData", sprintf("03_run_snpStats_workspace_CallFreq_%s_HWE_%s_MAF_%s.%s.RData", thres_call.freq, thres_hwe, thres_maf, group))
  
load(path.to.save.RData)
```

In this analysis, we use the package `snpStats` from `BioConductor` and `PLINK` to perform GWAS (Link: https://www.bioconductor.org/packages/release/bioc/html/snpStats.html, https://zzz.bwh.harvard.edu/plink/). 

Manhattan plot for p-values of all variants are generated by the package `regioneR` (https://bernatgel.github.io/karyoploter_tutorial//Tutorial/PlotManhattan/PlotManhattan.html)

# General information 
To perform ***Quality control**, we use the information from the `SNP_table`:

- `Call.Freq`: Call frequency. Remove variants whose `Call.Freq` is less than 50%; the variant exists in less than 50% of the samples.

- `ChiTest100`: Normalized Hardy-Weinberg p value calculated using genotype frequency. The value is calculated with 1 degree of freedom and normalized to 100 individuals. We remove variants that have `ChiTest100` <= 0.05.

- `Minor.Freq`: Minor allele frequency. We remove variants whose MAF is less than 0.01. 

Since we have two set of samples from two different plates, we need to merge the information from both sources.

# Quality control

## Distribution of the ChiTest100 p-value, cut-off at 0.05 {.tabset}

### Plate 1
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
ggplot(snp.full.info, aes(x = ChiTest100_plate1)) + 
    geom_histogram(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust=0.5, face="bold", size = 14)) +
    ylab("Cell density") +
    geom_vline(xintercept = thres_hwe, color = "red") +
    ggtitle("Distribution of Chi-squared HWE test, plate 1" )
```


### Plate 2 
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
ggplot(snp.full.info, aes(x = ChiTest100_plate2)) + 
    geom_histogram(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust=0.5, face="bold", size = 14)) +
    ylab("Cell density") +
    geom_vline(xintercept = thres_hwe, color = "red") +
    ggtitle("Distribution of Chi-squared HWE test, plate 2" )
```


## Distribution of the Call frequencies, cut-off at 50% {.tabset}

### Plate 1
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
ggplot(snp.full.info, aes(x = Call.Freq_plate1)) + 
    geom_histogram(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust=0.5, face="bold", size = 14)) +
    ylab("Cell density") +
    geom_vline(xintercept = thres_call.freq, color = "red") +
    ggtitle("Distribution of Call frequencies, plate 1")
```

### Plate 2
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
ggplot(snp.full.info, aes(x = Call.Freq_plate2)) + 
    geom_histogram(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust=0.5, face="bold", size = 14)) +
    ylab("Cell density") +
    geom_vline(xintercept = thres_call.freq, color = "red") +
    ggtitle("Distribution of Call frequencies, plate 2")
```


## Distribution of the Minor allele frequency (MAF), cut-off at 1% or 5% {.tabset}
### Plate 1
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
ggplot(snp.full.info, aes(x = Minor.Freq_plate1)) + 
    geom_histogram(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust=0.5, face="bold", size = 14)) +
    ylab("Cell density") +
    geom_vline(xintercept = thres_maf, color = "red") +
    ggtitle("Distribution of MAF, plate 1")
```

### Plate 2
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
ggplot(snp.full.info, aes(x = Minor.Freq_plate2)) + 
    geom_histogram(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust=0.5, face="bold", size = 14)) +
    ylab("Cell density") +
    geom_vline(xintercept = thres_maf, color = "red") +
    ggtitle("Distribution of MAF, plate 2")
```

## After filtering: 
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis'}
print(sprintf("Number of variants in the original dataset: %s", nrow(snp.full.info)))

filtered.snp.full.info <- subset(snp.full.info, QC.ChiTest100_plate1 == "PASS" & QC.ChiTest100_plate2 == "PASS" &
         QC.Call.Freq_plate1 == "PASS" & QC.Call.Freq_plate2 == "PASS" &
         QC.Minor.Freq_plate1 == "PASS" & QC.Minor.Freq_plate2 == "PASS")

print(sprintf("Number of variants in the filtered dataset: %s", nrow(filtered.snp.full.info)))

use.variants.in.GWAS <- filtered.snp.full.info$Name
snps.data <-  snps.data[, use.variants.in.GWAS]
snp.support <- subset(snp.support, snp.support$snp.names %in% use.variants.in.GWAS)
```


<!-- # Linkage disequilibrium statistics -->
<!-- ```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis'} -->
<!-- ``` -->


<!-- # Population structure (PCA) -->
<!-- This section is devoted to the Principal component analysis.  -->

# Descriptive stats.

## List of all samples
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis'} 
subject.support %>% rownames_to_column("Sample.ID") %>% subset(select = -c(member, pedigree, father, mother, sex)) %>% create_dt()
```

## Summary: Number of samples with and without disease/condition
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis'} 
subject.support %>% subset(select = -c(member, pedigree, father, mother, sex)) %>% 
  table() %>% 
  as.data.frame() %>%
  create_dt()
```

# PLINK RESULTS (with different model settings)
```{r echo=FALSE, warning=FALSE, message=FALSE, include=FALSE, results='asis', fig.width=14, fig.height=10}

path.to.plink.res <- file.path(outdir, PROJECT, "PLINK_results", sprintf("maf_%s_model", maf), sprintf("%s_with_models_results_maf_%s", group, maf), "plink.model")
plink.res <- read.table(path.to.plink.res, header = FALSE)
plink.res <- subset(plink.res, plink.res$V1 != "CHR")
plink.res <- plink.res %>% subset(select = c(V1, V2, V5, V6, V7, V8, V9, V10))
# plink.res <- subset(plink.res, plink.res$V1 != 0) %>%
  # subset(select = c(V1, V2, V3, V8, V9))

colnames(plink.res) <- c("chr", "name", "test", "AFF", "UNAFF", "chisq", "df", "pval")

plink.res <- merge(plink.res, snp.full.info, by.x = "name", by.y = "Name")
plink.res.original <- plink.res
plink.res <- subset(plink.res, QC.ChiTest100_plate1 == "PASS" & QC.ChiTest100_plate2 == "PASS" &
         QC.Call.Freq_plate1 == "PASS" & QC.Call.Freq_plate2 == "PASS" &
         QC.Minor.Freq_plate1 == "PASS" & QC.Minor.Freq_plate2 == "PASS")

plink.res <- subset(plink.res, plink.res$test == disease.model)
plink.res <- subset(plink.res, is.na(plink.res$pval) == FALSE)

```


## Visualization

### Manhattan plot
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=25, fig.height=12} 

plink.res$chr <- unlist(lapply(plink.res$Chr, function(x){
  return(sprintf("chr%s",x))
}))

row.names(plink.res) <- NULL

plink.res.to.grange <- subset(plink.res, select = c(chr, name, Position, pval)) %>%
  column_to_rownames("name")

plink.res.to.grange$end <- plink.res.to.grange$Position

plink.res.to.grange <- plink.res.to.grange[, c("chr", "Position", "end", "pval")]
plink.res.to.grange <- subset(plink.res.to.grange, plink.res.to.grange$pval != "NA")
plink.res.to.grange$pval <- as.numeric(plink.res.to.grange$pval)
colnames(plink.res.to.grange) <- c("chr", "start", "end", "pval")


plink.res.grange <- makeGRangesFromDataFrame(plink.res.to.grange, keep.extra.columns = TRUE)

library(karyoploteR)

kp <- plotKaryotype(plot.type=4)
kp <- kpPlotManhattan(kp, data= plink.res.grange)

plink.res$pval <- as.numeric(plink.res$pval) 

plink.pvaldf <- subset(plink.res, select = c(name, pval))
```


```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=25, fig.height=12} 
if ( nrow(subset(plink.res, plink.res$pval <= 5e-8 )) != 0){
  snps <- subset(plink.res, plink.res$pval <= 5e-8)$name
  # the next top 20
  snps.next20 <- plink.res %>% subset(pval > 5e-8) %>% arrange(desc(pval)) %>% tail(num.show.topSNPs)
  snps.next20 <- snps.next20$name
  snps <- unique(c(snps, snps.next20))
} else {
  snps <- plink.res %>% arrange(desc(pval)) %>% tail(num.show.topSNPs)
  snps <- snps$name
}

top.snps.plink <- kp$latest.plot$computed.values$data[snps]
```

### Manhattan plot with top SNPs

For better visualization, we just show top-20 SNPs. List of top-100 SNPs is given in the table below.

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=25, fig.height=12} 
kp <- plotKaryotype(plot.type=4)
kp <- kpPlotManhattan(kp, data=plink.res.grange, ymax=10)
kpAxis(kp, ymin = 0, ymax=10, numticks = 11)
kpText(kp, data = top.snps.plink %>% tail(20), labels = names(top.snps.plink %>% tail(20)), ymax=10, pos=3)
```


### Manhattan plot in each chromosome
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=30, fig.height=20} 
kp <- plotKaryotype(plot.type=1)
kpAxis(kp, ymin=0, ymax=10)
kp <- kpPlotManhattan(kp, data = plink.res.grange, points.col = "brewer.set3", ymax=10)
```

### Distribution of GWAS p-value
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
ggplot(plink.pvaldf, aes(x = pval)) + 
    geom_histogram(alpha = 0.2) +
    scale_x_log10() +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
    theme(plot.title = element_text(hjust=0.5, face="bold", size = 14)) +
    ylab("Cell density") +
    geom_vline(xintercept = 0.05, color = "red") +
    ggtitle("Distribution of GWAS P-VALUE")
```


# Downstream analysis

For the downstream analysis, we use only the output from `PLINK`. Note that ``PLINK` and `snpStats` might generate different results (different p-value) even if we have applied the same quality control steps. 

## List of candidate SNPs

Check if there is any SNPs that have `p-value < 5e-8`, otherwise show top-10 lowest p-value SNPs.

```{r echo=FALSE, include=FALSE}
create_dt <- function(x){
  DT::datatable(x,
                extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))
}

```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
top.snps.plink <- top.snps.plink %>% as.data.frame %>%  
  subset(select = c(seqnames, start, end, pval)) %>% 
  rownames_to_column("SNP.name") 
top.snps.plink %>% create_dt()
```

## MAF

In this section, we calculate the allele frequency of the second-most popular allele in the dataset (MAF, minor allele frequency) in the following three cases:

- All samples.

- Samples with a condition (e.g. with SBP, CDAD, ...). See the column `MAF.with.cond`.

- Samples without a condition (e.g. without SBP, without CDAD, ...). See the column `MAF.without.cond`.

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=10, fig.height=6}
##### HELPER FUNCTIONS
calculate_maf <- function(x){
  str_x <- str_split(paste(x, collapse = "/"), "/")[[1]]
  count.allele.A <- length( to_vec( for (item in str_x) if (item == "A") item ))
  count.allele.B <- length( to_vec( for (item in str_x) if (item == "B") item ))
  maf <- min(count.allele.A, count.allele.B)/length(str_x)
  return(maf)
}

candidate.snps <- peddf$genotypes[, top.snps.plink$SNP.name]
candidate.snps.df <- as(candidate.snps, "character") %>% as.data.frame() %>% rownames_to_column("Sample")
famdf <- peddf$fam %>% rownames_to_column("Sample") %>% subset(select = c(Sample, affected))
candidate.snps.df <- merge(candidate.snps.df, famdf, by.x = "Sample", by.y = "Sample")


all.maf <- c()
cond.maf <- c()
no.cond.maf <- c()
for (snp in top.snps.plink$SNP.name){
  tmpdf <- candidate.snps.df[, c("Sample", "affected", snp)]
  tmpdf <- subset(tmpdf, tmpdf[, snp] != "NA")
  alleles.cond <- subset(tmpdf, tmpdf$affected == 2)
  alleles.no.cond <- subset(tmpdf, tmpdf$affected == 1)
  all.maf <- c(all.maf, calculate_maf(tmpdf[, snp]))
  cond.maf <- c(cond.maf, calculate_maf(alleles.cond[, snp]))
  no.cond.maf <- c(no.cond.maf, calculate_maf(alleles.no.cond[, snp]))
}
top.snps.plink$MAF <- all.maf
top.snps.plink$MAF.with.cond <- cond.maf
top.snps.plink$MAF.without.cond <- no.cond.maf

top.snps.plink %>% mutate_if(is.numeric, round, 6) %>% create_dt()
```

## Odd ratios {.tabset}

We calculate the **odd ratios** and summarize all information in the following tables:

- `genotype`: all genotypes occurring in the data at the given SNP (AA, AB, BB).

- `without.cond`: Number of samples which have the given `genotype` and do not have the "disease/condition" (e.g. SBP, ...). 

- `with.cond`: Number of samples which have the given `genotype` and have the "disease/condition" (e.g. SBP, ...).

- `SNP`: name of the SNP.

- `total`: total number of samples having the given `genotype`.

- `genotype.type`: this could be *reference.gt* (the most dominant genotype in the dataset), or *genotype.1* and *genotype.2*.

- `odd`: disease/condition odd for each genotype. Calulated by: `(with.cond / total) / (1 - with.cond/total)`.

- `odd.ratio`: odd ratio, calculate by divided the given genotype's odd by the reference genotype's odd.

- `with.cond.frequency`: frequency of samples which have the disease/condition (e.g. SBP) with a given `genotype`. Calculated by `with.cond/total`. 

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', include=FALSE, fig.width=10, fig.height=6}
candidate.snps.df[1:2, 1:2] %>% create_dt()
```

If the number of candidate SNPs (p-value <= 5e-8) is larger than 20, we show all SNPs in a single table, please use the search box on top of the column `SNP` to filter the SNP. 

If the number of candidate SNPs (p-value <= 5e-8) is smaller than 20 or we just show top-10 lowest p-value SNPs, each SNP will be displayed in a separated tab. 

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', include=TRUE, fig.width=10, fig.height=6}
all.countdf <- data.frame()

if (length(top.snps.plink$SNP.name) <= 20){
  for (snp in top.snps.plink$SNP.name){
  
    tmpdf <- candidate.snps.df[, c("Sample", "affected", snp)]
    tmpdf <- subset(tmpdf, tmpdf[, snp] != "NA")
    
    countdf <- tmpdf[, c(snp, "affected")] %>% table() %>% as.data.frame() %>% pivot_wider(names_from = "affected", values_from = "Freq")
    
    if (2 %in% colnames(countdf) == FALSE){
      countdf$`2` <- c(0)
    } else if (1 %in% colnames(countdf) == FALSE){
      countdf$`1` <- c(0)
    }
   
    colnames(countdf) <- c("genotype", "without.cond", "with.cond")
    countdf$SNP <- c(snp)
    countdf <- countdf %>% rowwise() %>% mutate(total = without.cond + with.cond) %>% arrange(desc(total))
    
    countdf$genotype.type <- c("reference.gt", to_vec(for (item in seq(1, nrow(countdf) - 1)) sprintf("genotype.%s", item)))
    
    countdf <- countdf %>% rowwise() %>% 
      mutate(odd = (with.cond / (with.cond + without.cond)) / (1 - (with.cond / (with.cond + without.cond))))
    
    ref.odd <- subset(countdf, countdf$genotype.type == "reference.gt")$odd
    
    countdf <- countdf %>% rowwise() %>%  
      mutate(odd.ratio = odd / ref.odd) %>%
      mutate(with.cond.frequency = with.cond / total)
    
    cat(paste("\n\n### ", snp, "\n"))
    print( htmltools::tagList(datatable(countdf %>% mutate_if(is.numeric, round, 6), extensions = 'Buttons',
                  filter = "top",
                  options = list(dom = 'Blfrtip',
                                 buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                                 lengthMenu = list(c(10,25,50,-1),
                                                   c(10,25,50,"All")),
                                 columnDefs = list(list(
                                   targets = "_all",
                                   render = JS(
                                     "function(data, type, row, meta) {",
                                     "return type === 'display' && data != null && data.length > 100 ?",
                                     "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                     "}")
                                 ))
                  ))))
    cat("\n \n")
  }
} else {
    for (snp in top.snps.plink$SNP.name){
    
    tmpdf <- candidate.snps.df[, c("Sample", "affected", snp)]
    tmpdf <- subset(tmpdf, tmpdf[, snp] != "NA")
    
    countdf <- tmpdf[, c(snp, "affected")] %>% table() %>% as.data.frame() %>% pivot_wider(names_from = "affected", values_from = "Freq")
    
    if (2 %in% colnames(countdf) == FALSE){
      countdf$`2` <- c(0)
    } else if (1 %in% colnames(countdf) == FALSE){
      countdf$`1` <- c(0)
    }
   
    colnames(countdf) <- c("genotype", "without.cond", "with.cond")
    countdf$SNP <- c(snp)
    countdf <- countdf %>% rowwise() %>% mutate(total = without.cond + with.cond) %>% arrange(desc(total))
    
    countdf$genotype.type <- c("reference.gt", to_vec(for (item in seq(1, nrow(countdf) - 1)) sprintf("genotype.%s", item)))
    
    countdf <- countdf %>% rowwise() %>% 
      mutate(odd = (with.cond / (with.cond + without.cond)) / (1 - (with.cond / (with.cond + without.cond)))) %>%
      mutate(odd = round(odd, 6))
    
    ref.odd <- subset(countdf, countdf$genotype.type == "reference.gt")$odd
    
    countdf <- countdf %>% rowwise() %>%  
      mutate(odd.ratio = round(odd / ref.odd, 6)) %>%
      mutate(with.cond.frequency = round(with.cond / total, 6))
    all.countdf <- rbind(all.countdf, countdf)
    }
  all.countdf %>% create_dt()
}
```

# Candidate regions to perform postGWAS

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', include=TRUE, fig.width=10, fig.height=6}
path.to.save.candidate.regions <- file.path(outdir, PROJECT, "candidate_regions", group, sprintf("MAF_%s", maf))
dir.create(path.to.save.candidate.regions, showWarnings = FALSE, recursive = TRUE)
candidate.regiondf <- tail(top.snps.plink, 20) %>% subset(select = c(seqnames, start)) %>%
  rowwise() %>%
  mutate(end = start + 1e6) %>%
  mutate(start = ifelse(start - 1e6 >= 0, start - 1e6, 0) ) %>%
  mutate(region.length = end - start)

write.table(candidate.regiondf, file.path(path.to.save.candidate.regions, sprintf("%s_MAF_%s_model_%s.tsv", group, maf, disease.model)), sep = "\t", col.names = FALSE, row.names = FALSE)

candidate.regiondf %>% create_dt()
```

